aten/src/ATen/Context.h

```diff
@@ -56,14 +57,24 @@ class CAFFE2_API Context {
       LegacyTHDispatch::LegacyTHDispatcherUniquePtr{t, LegacyTHDispatcherDeleter([](LegacyTHDispatcher* p) { delete p; }) });
   }
 
-  Generator & defaultGenerator(DeviceType device_type) {
-    initCUDAIfNeeded(device_type);
-    initHIPIfNeeded(device_type);
-    auto & generator = generator_registry[static_cast<int>(device_type)];
-    if(!generator)
-      AT_ERROR(DeviceTypeName(device_type), " backend type not enabled.");
-    return *generator;
+  // PyTorch maintains a collection of default generators that get
+  // inialized once. The purpose of these default generator is to
+  // maintain a running state of the pseudo random number generation. 
+  // getDefaultGenerator gets the default generator for a particular
+  // device
```

Comment:
So, eventually, are we going to put CPUGenerator into the generator registry? We certainly can't keep doing the current strategy (if-statements) when CUDA rolls around, because you won't be allowed to access that symbol from here.

Reply:
So the problem with generator_registry is, CUDAGenerator will have its own std::vector of default generators, where the size of std::vector is the number of GPUs in the machine. Currently size of generator_registry is the number of backends in PyTorch. Hence,  For the symbols, I was planning on the doing the if statements and call at::getCUDAHooks().getCUDAGenerator(device). Wouldn't that work?

Reply:
That would work, I suppose, but in that case we might as well just give up on the generator registry and do this entirely through hooks.